generators/chipyard/src/main/scala/GenericParams.scala

```diff
@@ -0,0 +1,177 @@
+package chipyard
+
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe._
+
+import chisel3._
+
+import freechips.rocketchip.config.{Parameters, Config, Field, View}
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.diplomacy.{LazyModule, ClockCrossingType, ValName}
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree.LogicalTreeNode
+import freechips.rocketchip.rocket._
+import freechips.rocketchip.tile._
+
+// Trait for generic case class of base trait for copying
+trait ConcreteBaseTrait[Base] {
+  this: Product =>
+  val _origin: Base
+
+  // Convert back to core-specific tile
+  def convert: Base = {
+    // Reflection Info of this class
+    val fieldNames = (this.getClass.getDeclaredFields map (f => f.getName)).init
+
+    // Reflection of target class
+    val paramClass = _origin.getClass
+    val paramNames = (paramClass.getDeclaredFields map (f => f.getName))
+    val paramCtor = paramClass.getConstructors.head
+
+    // Build a list of parameter in the original parameter class
+    val nameDict = paramNames.zipWithIndex.toMap
+    val indexList = fieldNames map (n => nameDict.get(n))
+    val fieldList = this.productIterator.toList map {
+      case c: ConcreteBaseTrait[_] => c.convert
+      case v => v
+    }
+    val fieldDict = ((indexList zip fieldList) collect { case (Some(i), v) => (i, v) }).toMap
+    val newValues = _origin.asInstanceOf[Product].productIterator.toList.zipWithIndex map 
+      { case (v, i) => (if (fieldDict contains i) fieldDict(i) else v).asInstanceOf[AnyRef] }
+
+    paramCtor.newInstance(newValues:_*).asInstanceOf[Base]
+  }
+}
+
+// Case class to change common parameters visible in the base traits. Some fields in the base traits may not be configurable as a 
+// case class constructor parameter for some cores, and those field will be ignored when applied.
+case class GenericCoreParams(
+  val bootFreqHz: BigInt,
+  val useVM: Boolean,
+  val useUser: Boolean,
+  val useSupervisor: Boolean,
+  val useDebug: Boolean,
+  val useAtomics: Boolean,
+  val useAtomicsOnlyForIO: Boolean,
+  val useCompressed: Boolean,
+  override val useVector: Boolean,
+  val useSCIE: Boolean,
+  val useRVE: Boolean,
+  val mulDiv: Option[MulDivParams],
+  val fpu: Option[FPUParams],
+  val fetchWidth: Int,
+  val decodeWidth: Int,
+  val retireWidth: Int,
+  val instBits: Int,
+  val nLocalInterrupts: Int,
+  val nPMPs: Int,
+  val pmpGranularity: Int,
+  val nBreakpoints: Int,
+  val useBPWatch: Boolean,
+  val nPerfCounters: Int,
```

Comment:
I'm not sure this is a good idea. This object seems to attempt to create a superset of all possible core parameters, which is not a sustainable approach,

Reply:
This class is intended to provide a `copy()` function only for the core parameters in `TileParams` or `CoreParams`, which would be used by all cores. Chipyard is not supposed to change non-standard, core-specific parameters with generic config fragments since we don't even know what they means for different cores.

However, if the user want to create their own generic config fragments for their family of cores which contains some shared parameters not included in the standard set of parameter classes, then they can put those shared parameters in a trait and use the similar technique in this file to implement them. 

For example, there is a config fragment that enables trace port (`WithTraceIO`) in Chipyard. The parameter `trace` is not included in `TileParams` and is not used by Rocket, but it is intended to be generic for at least the cores that support toggleable trace ports generation, and it is used by other general purpose config fragments in FireSim. A new trait containing `trace` only and inherited by Boom and Ariane (which support toggleable trace port generation) can be created along with a generic case class to generalize this config fragment.